File: 树状数组/130BIT.cpp
#include <iostream>
#include <vector>

using namespace std;

typedef long long ll;

#define MAX_SIZE 1000007
ll seq[MAX_SIZE];
ll tree[MAX_SIZE];

class BIT {
public:
    ll n;

    int lowbit(int x) {
        return x & (-x);
    }

    BIT(ll n, ll * seq) : n(n) {
        for (int i = 1; i <= n; ++i) {
            tree[i] += seq[i-1];
            int j = i + lowbit(i);
            if (j <= n) {
                tree[j] += tree[i];
            }
        }
    }

    void add(int i, ll x) {
        while (i <= n) {
            tree[i] += x;
            i += lowbit(i);
        }
    }

    ll query(int i) {
        ll ret = 0;
        while (i) {
            ret += tree[i];
            i -= lowbit(i);
        }
        return ret;
    }

    ll query(int l, int r) {
        return query(r) - query(l - 1);
    }
};

int main(void) {
    ll n, q;
    cin >> n >> q;
    for (int i = 0; i < n; ++i) {
        cin >> seq[i];
    }
    BIT bit(n, seq);
    while (q) {
        ll op, x, y;
        cin >> op >> x >> y;
        if (op == 1) {
            bit.add(x, y);
        } else {
            cout << bit.query(x, y) << endl;
        }
        q -= 1;
    }
    return 0;
}

File: 树状数组/130BIT.py
from typing import List
import collections
import math
from functools import lru_cache
import queue
import bisect
import heapq
import copy
import sys

class BIT:
    def lowbit(self, x):
        return x & (-x)

    def __init__(self, seq):
        self.n = len(seq)
        self.t = [0] * (n+1)
        for x, e in enumerate(seq):
            x += 1
            self.t[x] += e
            j = x + self.lowbit(x)
            if j <= n:
                self.t[j] += self.t[x]

    def add(self, i, x):
        while i <= self.n:
            self.t[i] += x
            i += self.lowbit(i)

    def find(self, i):
        ret = 0
        while i:
            ret += self.t[i]
            i -= self.lowbit(i)

        return ret

    def query(self, l, r):
        return self.find(r) - self.find(l-1)


if __name__ == '__main__':
    line = sys.stdin.readline()[:-1].split(" ")
    n, q = int(line[0]), int(line[1])
    line = sys.stdin.readline()[:-1].split(" ")
    seq = [int(e) for e in line]
    bit = BIT(seq)

    for _ in range(q):
        line = sys.stdin.readline()[:-1].split(" ")
        op, x, y = int(line[0]), int(line[1]), int(line[2])
        if op == 1:
            bit.add(x, y)
        else:
            print(bit.query(x, y))
